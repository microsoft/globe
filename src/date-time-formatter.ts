/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { enUS } from "date-fns/locale";
import { CachedDateTimeFormat } from "./cached-datetimeformat";
import {
  DateTimeFormatOptions,
  FULL,
  FULL_DATE,
  FULL_DATE_WITH_YEAR,
  FULL_TIME,
  FULL_WITH_YEAR,
  HOUR_ONLY,
  LONG_DATE,
  LONG_DATE_WITH_YEAR,
  LONG_TIME,
  LONG_TIME_WITH_TIMEZONE,
  LONG_WEEKDAY,
  LONG_WEEKDAY_LONG_TIME,
  LONG_WEEKDAY_SHORT_TIME,
  LONG_WITH_TIMEZONE,
  LONG_WITH_YEAR_TIMEZONE,
  MEDIUM,
  MEDIUM_DATE,
  MEDIUM_DATE_SHORT_TIME,
  MEDIUM_DATE_WITH_YEAR,
  MEDIUM_TIME,
  MEDIUM_WITH_YEAR,
  SHORT,
  SHORT_DATE,
  SHORT_DATE_LONG_TIME,
  SHORT_DATE_TIME,
  SHORT_DATE_WITH_SHORT_YEAR,
  SHORT_DATE_WITH_YEAR,
  SHORT_TIME,
  SHORT_WEEKDAY,
  SHORT_WEEKDAY_LONG_TIME,
  SHORT_WEEKDAY_SHORT_TIME,
  SHORT_WITH_YEAR,
} from "./date-time-format-options";
import ILocaleInfo from "./ILocaleInfo";
import { OsDateTimeFormatter } from "./os-date-time-formatter";

export class DateTimeFormatter {
  // We're keying this using JSON.stringify because with a WeakMap we've have a key pair
  // (locale - string & options - object) and stringify is native so it is so fars it is
  // not worth maintaing the two-level cache (map for string and weak map for object)
  private readonly cachedDateTimeFormat = new CachedDateTimeFormat();
  private dateFnsLocale: any = enUS;
  private formatter?: OsDateTimeFormatter = undefined;

  /**
   * Instantiates DateTimeFormatter
   * @param locale The desired locale to which to format the date and time value (default: en-US)
   */
  constructor(private locale: string | ILocaleInfo = "en-US") {
    console.log("constructor");
    if (typeof this.locale === "string") {
      this.loadLocale(this.locale);
    } else {
      this.loadLocale(this.locale.regionalFormat);
    }
  }

  private loadLocale(locale: string) {
    const fnsLocale = locale.toLocaleLowerCase().split("-")[0];
    console.log(`fnsLocale: ${fnsLocale}`);
    if (fnsLocale === "en") {
      return;
    }
    try {
      console.log(`Loading date-fns locale: ${fnsLocale}`);
      /* webpackMode: "lazy", webpackChunkName: "df-[index]", webpackExclude: /_lib/ */
      import(`date-fns/locale/${fnsLocale}`).then(locale => {
        this.dateFnsLocale = locale;
        console.log("Successfully loaded date-fns locale", this.dateFnsLocale);
      });
    } catch (e) {
      this.dateFnsLocale = enUS;
      console.error(`Failed to load date-fns locale: ${locale}`);
    }
  }

  /**
   * Localizes the date/time value
   * @param date The date/time to localize
   * @param format The format to be used for the localization
   * @returns The localized date/time string
   */
  public formatDateTime(date: number | Date, format: DateTimeFormatOptions) {
    if (typeof this.locale === "string") {
      const dtf = this.cachedDateTimeFormat.get(this.locale, format);
      return dtf.format(date);
    }

    return this.formatOsDateTime(date, format, this.locale);
  }

  public formatOsDateTime(
    date: number | Date,
    format: DateTimeFormatOptions,
    localeInfo: ILocaleInfo
  ): string {
    if (!localeInfo) {
      throw new Error(
        "Cannot call the OS date and time formatter without specifying the OS locale info."
      );
    }

    if (!this.formatter) {
      this.formatter = new OsDateTimeFormatter(
        localeInfo.regionalFormat,
        localeInfo.platform,
        this.cachedDateTimeFormat,
        this.dateFnsLocale
      );
    }

    let loc: string;
    if (typeof this.locale === "string") {
      loc = this.locale;
    } else {
      loc = this.locale.regionalFormat;
    }

    switch (format) {
      case SHORT_TIME: {
        if (!localeInfo.shortTime) {
          throw new Error(`localeInfo.shortTime was not provided!`);
        }

        return this.formatter.timeToString(date, localeInfo.shortTime);
      }
      case SHORT_DATE:
      case SHORT_DATE_WITH_SHORT_YEAR:
      case SHORT_DATE_WITH_YEAR: {
        if (!localeInfo.shortDate) {
          throw new Error(`localeInfo.shortDate was not provided!`);
        }

        return this.formatter.dateToString(date, localeInfo.shortDate);
      }
      case SHORT_DATE_LONG_TIME: {
        if (!localeInfo.shortDate || !localeInfo.longTime) {
          throw new Error(
            `localeInfo.shortDate or localeInfo.longTime was not provided!`
          );
        }

        const d = this.formatter.dateToString(date, localeInfo.shortDate);
        const t = this.formatter.timeToString(date, localeInfo.longTime);
        return this.combineDateAndTime(d, t);
      }
      case SHORT:
      case SHORT_WITH_YEAR:
      case SHORT_DATE_TIME: {
        if (!localeInfo.shortDate || !localeInfo.shortTime) {
          throw new Error(
            `localeInfo.shortDate or localeInfo.shortTime was not provided!`
          );
        }

        const d = this.formatter.dateToString(date, localeInfo.shortDate);
        const t = this.formatter.timeToString(date, localeInfo.shortTime);
        return this.combineDateAndTime(d, t);
      }
      case HOUR_ONLY: {
        if (!localeInfo.shortTime) {
          throw new Error(`localeInfo.shortTime was not provided!`);
        }

        if (localeInfo.platform === "macos") {
          const includesADayPeriod = localeInfo.shortTime.includes("a");
          return this.formatter.timeToString(
            date,
            includesADayPeriod ? "h a" : "H"
          );
        } else {
          const includesTTDayPeriod = localeInfo.shortTime.includes("tt");
          const includesTDayPeriod = localeInfo.shortTime.includes("t");
          return this.formatter.timeToString(
            date,
            includesTTDayPeriod ? "h tt" : includesTDayPeriod ? "h t" : "H"
          );
        }
      }
      case FULL_TIME:
      case MEDIUM_TIME:
      case LONG_TIME:
      case LONG_TIME_WITH_TIMEZONE: {
        if (!localeInfo.longTime) {
          throw new Error(`localeInfo.longTime was not provided!`);
        }

        const time = this.formatter.timeToString(date, localeInfo.longTime);
        return format === LONG_TIME_WITH_TIMEZONE || format === FULL_TIME
          ? this.ensureTimeZone(
              time,
              date,
              localeInfo.longTime,
              format,
              localeInfo
            )
          : time;
      }
      case MEDIUM:
      case MEDIUM_WITH_YEAR: {
        if (!localeInfo.longDate || !localeInfo.longTime) {
          throw new Error(
            `localeInfo.longDate or localeInfo.longTime was not provided!`
          );
        }

        const d = this.formatter.dateToString(date, localeInfo.longDate);
        const t = this.formatter.timeToString(date, localeInfo.longTime);
        return this.combineDateAndTime(d, t);
      }
      case MEDIUM_DATE_SHORT_TIME: {
        if (!localeInfo.shortTime) {
          throw new Error(`localeInfo.shortTime was not provided!`);
        }

        const d = this.cachedDateTimeFormat.get(loc, MEDIUM_DATE).format(date);
        const t = this.formatter.timeToString(date, localeInfo.shortTime);

        return `${d}, ${t}`;
      }
      case FULL:
      case FULL_WITH_YEAR: {
        let dateFormat = localeInfo.longDate;
        if (localeInfo.fullDate && localeInfo.fullDate !== "UNKNOWN") {
          dateFormat = localeInfo.fullDate;
        }

        if (!dateFormat || !localeInfo.longTime) {
          throw new Error(
            `localeInfo.longDate or localeInfo.longTime was not provided!`
          );
        }

        const d = this.formatter.dateToString(date, dateFormat);
        const t = this.formatter.timeToString(date, localeInfo.longTime);
        const timeWithTimeZone = this.ensureTimeZone(
          t,
          date,
          localeInfo.longTime,
          format,
          localeInfo
        );
        return this.combineDateAndTime(d, timeWithTimeZone);
      }
      case LONG_WEEKDAY:
      case SHORT_WEEKDAY: {
        return this.cachedDateTimeFormat.get(loc, format).format(date);
      }
      case LONG_WEEKDAY_LONG_TIME:
      case SHORT_WEEKDAY_LONG_TIME: {
        if (!localeInfo.longDate || !localeInfo.longTime) {
          throw new Error(
            `localeInfo.longDate or localeInfo.longTime was not provided!`
          );
        }

        const weekFormat =
          format === LONG_WEEKDAY_LONG_TIME
            ? this.cachedDateTimeFormat.get(loc, LONG_WEEKDAY).format(date)
            : this.cachedDateTimeFormat.get(loc, SHORT_WEEKDAY).format(date);

        return `${weekFormat}, ${this.formatter.timeToString(
          date,
          localeInfo.longTime
        )}`;
      }
      case LONG_WEEKDAY_SHORT_TIME:
      case SHORT_WEEKDAY_SHORT_TIME: {
        if (!localeInfo.shortTime) {
          throw new Error(`localeInfo.shortTime was not provided!`);
        }

        const weekFormat =
          format === LONG_WEEKDAY_SHORT_TIME
            ? this.cachedDateTimeFormat.get(loc, LONG_WEEKDAY).format(date)
            : this.cachedDateTimeFormat.get(loc, SHORT_WEEKDAY).format(date);

        return `${weekFormat}, ${this.formatter.timeToString(
          date,
          localeInfo.shortTime
        )}`;
      }
      case MEDIUM_DATE:
      case LONG_DATE:
      case MEDIUM_DATE_WITH_YEAR:
      case LONG_DATE_WITH_YEAR: {
        if (!localeInfo.longDate) {
          throw new Error(`localeInfo.longDate was not provided!`);
        }

        return this.formatter.dateToString(date, localeInfo.longDate);
      }

      case FULL_DATE:
      case FULL_DATE_WITH_YEAR: {
        if (localeInfo.fullDate && localeInfo.fullDate !== "UNKNOWN") {
          return this.formatter.dateToString(date, localeInfo.fullDate);
        }

        if (!localeInfo.longDate) {
          throw new Error(`localeInfo.longDate was not provided!`);
        }

        return this.formatter.dateToString(date, localeInfo.longDate);
      }

      case LONG_WITH_TIMEZONE:
      case LONG_WITH_YEAR_TIMEZONE: {
        if (!localeInfo.longDate || !localeInfo.longTime) {
          throw new Error(
            `localeInfo.longDate or localeInfo.longTime was not provided!`
          );
        }

        const d = this.formatter.dateToString(date, localeInfo.longDate);
        const t = this.formatter.timeToString(date, localeInfo.longTime);
        const timeWithTimeZone = this.ensureTimeZone(
          t,
          date,
          localeInfo.longTime,
          format,
          localeInfo
        );
        return this.combineDateAndTime(d, timeWithTimeZone);
      }
    }

    let formatStringified = undefined;
    try {
      formatStringified = JSON.stringify(format);
      // eslint-disable-next-line no-empty
    } catch {}

    throw new Error(
      "Incorrect OS locale info format specified:" +
        (formatStringified || format)
    );
  }

  private combineDateAndTime(date: string, time: string) {
    return `${date} ${time}`;
  }

  private ensureTimeZone(
    time: string,
    date: number | Date,
    mask: string,
    format: Intl.DateTimeFormatOptions,
    localeInfo: ILocaleInfo
  ) {
    if (this.formatter?.timeHasTimeZone(mask)) {
      return time;
    }

    const timeZoneName = this.formatter?.getTimeZoneName(date, format);
    return `${time} ${timeZoneName}`;
  }
}
