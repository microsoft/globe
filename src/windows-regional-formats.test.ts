/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { describe, it, expect } from "vitest";
import {
  FULL,
  FULL_DATE,
  FULL_DATE_WITH_YEAR,
  FULL_TIME,
  FULL_WITH_YEAR,
  HOUR_ONLY,
  LONG_DATE,
  LONG_DATE_WITH_YEAR,
  LONG_TIME,
  LONG_TIME_WITH_TIMEZONE,
  LONG_WEEKDAY,
  LONG_WEEKDAY_LONG_TIME,
  LONG_WEEKDAY_SHORT_TIME,
  LONG_WITH_TIMEZONE,
  LONG_WITH_YEAR_TIMEZONE,
  MEDIUM,
  MEDIUM_DATE,
  MEDIUM_DATE_SHORT_TIME,
  MEDIUM_DATE_WITH_YEAR,
  MEDIUM_TIME,
  MEDIUM_WITH_YEAR,
  SHORT,
  SHORT_DATE,
  SHORT_DATE_LONG_TIME,
  SHORT_DATE_TIME,
  SHORT_DATE_WITH_SHORT_YEAR,
  SHORT_DATE_WITH_YEAR,
  SHORT_TIME,
  SHORT_WEEKDAY,
  SHORT_WEEKDAY_LONG_TIME,
  SHORT_WEEKDAY_SHORT_TIME,
  SHORT_WITH_YEAR,
} from "./date-time-format-options";
import { DateTimeFormatter } from "./date-time-formatter";
import { RegionalSettingsPerLocale } from "./windows-regional-formats";

describe("Windows Regional Formats", () => {
  // const supportedLocales = [
  //   "ar-sa",
  //   "az-latn-az",
  //   "bg-bg",
  //   "ca-es",
  //   "ca-es-valencia",
  //   "cs-cz",
  //   "cy-gb",
  //   "da-dk",
  //   "de-at",
  //   "de-de",
  //   "de-li",
  //   "de-lu",
  //   "de-ch",
  //   "nl-be",
  //   "nl-nl",
  //   "en-au",
  //   "en-bz",
  //   "en-ca",
  //   "en-029",
  //   "en-in",
  //   "en-ie",
  //   "en-jm",
  //   "en-my",
  //   "en-nz",
  //   "en-ph",
  //   "en-sg",
  //   "en-za",
  //   "en-tt",
  //   "en-gb",
  //   "en-us",
  //   "en-zw",
  //   "et-ee",
  //   "fi-fi",
  //   "fil-ph",
  //   "fr-be",
  //   "fr-ca",
  //   "fr-fr",
  //   "fr-lu",
  //   "fr-mc",
  //   "fr-ch",
  //   "gl-es",
  //   "ka-ge",
  //   "he-il",
  //   "hi-in",
  //   "hr-hr",
  //   "hu-hu",
  //   "is-is",
  //   "id-id",
  //   "it-it",
  //   "it-ch",
  //   "ja-jp",
  //   "kk-kz",
  //   "ko-kr",
  //   "lv-lv",
  //   "lt-lt",
  //   "mk-mk",
  //   "nb-no",
  //   "nn-no",
  //   "pl-pl",
  //   "pt-br",
  //   "pt-pt",
  //   "ro-ro",
  //   "ru-ru",
  //   "sq-al",
  //   "sr-latn-rs",
  //   "sk-sk",
  //   "sl-si",
  //   "es-ar",
  //   "es-ve",
  //   "es-bo",
  //   "es-cl",
  //   "es-co",
  //   "es-cr",
  //   "es-do",
  //   "es-ec",
  //   "es-sv",
  //   "es-gt",
  //   "es-hn",
  //   "es-mx",
  //   "es-ni",
  //   "es-py",
  //   "es-pa",
  //   "es-pe",
  //   "es-pr",
  //   "es-es",
  //   "es-us",
  //   "es-uy",
  //   "sv-fi",
  //   "sv-se",
  //   "th-th",
  //   "tr-tr",
  //   "uk-ua",
  //   "vi-vn",
  //   "zh-cn",
  //   "zh-sg",
  //   "zh-hk",
  //   "zh-mo",
  //   "zh-tw",
  // ];
  const supportedLocales = ["es-es", "ro-ro"];
  const formatsToTest: { [key: string]: any } = {
    SHORT,
    SHORT_WITH_YEAR,
    SHORT_TIME,
    SHORT_DATE,
    SHORT_DATE_WITH_YEAR,
    SHORT_DATE_WITH_SHORT_YEAR,
    SHORT_DATE_TIME,
    LONG_DATE,
    LONG_TIME_WITH_TIMEZONE,
    LONG_WITH_TIMEZONE,
    LONG_WITH_YEAR_TIMEZONE,
    SHORT_DATE_LONG_TIME,
    LONG_DATE_WITH_YEAR,
    LONG_TIME,
    LONG_WEEKDAY,
    SHORT_WEEKDAY,
    FULL_DATE_WITH_YEAR,
    FULL_DATE,
    FULL_TIME,
    FULL_WITH_YEAR,
    FULL,
    MEDIUM_TIME,
    MEDIUM_DATE,
    MEDIUM_DATE_WITH_YEAR,
    MEDIUM_WITH_YEAR,
    MEDIUM,
    MEDIUM_DATE_SHORT_TIME,
    HOUR_ONLY,
    LONG_WEEKDAY_SHORT_TIME,
    LONG_WEEKDAY_LONG_TIME,
    SHORT_WEEKDAY_SHORT_TIME,
    SHORT_WEEKDAY_LONG_TIME,
  };
  const date = new Date(2020, 0, 15, 13, 30, 45);
  supportedLocales.forEach(locale => {
    const regionalFormat = RegionalSettingsPerLocale.filter(
      regionalFormat =>
        regionalFormat.Culture.toLocaleLowerCase() ===
        locale.toLocaleLowerCase()
    )[0];

    // let localeBasedDateTimeFormat;
    let regionalBasedDateTimeFormatter: DateTimeFormatter | undefined;
    // try {
    //   const localeBasedDateTimeFormatter = new DateTimeFormatter(locale);
    //   localeBasedDateTimeFormat = localeBasedDateTimeFormatter.formatDateTime(
    //     date,
    //     format
    //   );
    // } catch {
    //   localeBasedDateTimeFormat = "Not supported";
    // }

    try {
      regionalBasedDateTimeFormatter = new DateTimeFormatter({
        platform: "windows",
        regionalFormat: regionalFormat.Culture,
        shortDate: regionalFormat.ShortDateFormat,
        longDate: regionalFormat.LongDateFormat,
        shortTime: regionalFormat.ShortTimeFormat,
        longTime: regionalFormat.LongTimeFormat,
      });
    } catch {
      console.log(`Regional settings not supported for ${locale}`);
    }
    // it(`Locale based formatting ${locale} - ${formatName}`, () => {
    //   expect(localeBasedDateTimeFormat).toMatchSnapshot();
    // });

    Object.keys(formatsToTest).forEach((formatName: string) => {
      const format = formatsToTest[formatName];

      it(`Regional settings based formatting ${locale} - ${formatName}`, () => {
        const regionalBasedDateTimeFormat = regionalBasedDateTimeFormatter
          ? regionalBasedDateTimeFormatter.formatDateTime(date, format)
          : "Not supported";
        expect(regionalBasedDateTimeFormat).toMatchSnapshot();
      });
    });
  });
});
