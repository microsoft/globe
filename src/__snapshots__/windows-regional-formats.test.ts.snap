// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Windows Regional Formats > Regional settings based formatting en-us - FULL 1`] = `"Wednesday, January 15, 2020 13:30:45 01/15/20201 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - FULL_DATE 1`] = `"Wednesday, January 15, 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - FULL_DATE_WITH_YEAR 1`] = `"Wednesday, January 15, 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - FULL_TIME 1`] = `"13:30:45 01/15/20201 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - FULL_WITH_YEAR 1`] = `"Wednesday, January 15, 2020 13:30:45 01/15/20201 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - HOUR_ONLY 1`] = `"13 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_DATE 1`] = `"Wednesday, January 15, 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_DATE_WITH_YEAR 1`] = `"Wednesday, January 15, 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_TIME 1`] = `"13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_TIME_WITH_TIMEZONE 1`] = `"13:30:45 01/15/20201 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_WEEKDAY 1`] = `"Wednesday"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_WEEKDAY_LONG_TIME 1`] = `"Wednesday, 13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_WEEKDAY_SHORT_TIME 1`] = `"Wednesday, 13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_WITH_TIMEZONE 1`] = `"Wednesday, January 15, 2020 13:30:45 01/15/20201 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - LONG_WITH_YEAR_TIMEZONE 1`] = `"Wednesday, January 15, 2020 13:30:45 01/15/20201 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - MEDIUM 1`] = `"Wednesday, January 15, 2020 13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - MEDIUM_DATE 1`] = `"Wednesday, January 15, 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - MEDIUM_DATE_SHORT_TIME 1`] = `"Jan 15, 13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - MEDIUM_DATE_WITH_YEAR 1`] = `"Wednesday, January 15, 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - MEDIUM_TIME 1`] = `"13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - MEDIUM_WITH_YEAR 1`] = `"Wednesday, January 15, 2020 13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT 1`] = `"1/15/2020 13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_DATE 1`] = `"1/15/2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_DATE_LONG_TIME 1`] = `"1/15/2020 13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_DATE_TIME 1`] = `"1/15/2020 13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_DATE_WITH_SHORT_YEAR 1`] = `"1/15/2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_DATE_WITH_YEAR 1`] = `"1/15/2020"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_TIME 1`] = `"13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_WEEKDAY 1`] = `"Wed"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_WEEKDAY_LONG_TIME 1`] = `"Wed, 13:30:45 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_WEEKDAY_SHORT_TIME 1`] = `"Wed, 13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting en-us - SHORT_WITH_YEAR 1`] = `"1/15/2020 13:30 01/15/20201"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - FULL 1`] = `"miércoles, 15 de enero de 2020 13:30:45 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - FULL_DATE 1`] = `"miércoles, 15 de enero de 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - FULL_DATE_WITH_YEAR 1`] = `"miércoles, 15 de enero de 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - FULL_TIME 1`] = `"13:30:45 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - FULL_WITH_YEAR 1`] = `"miércoles, 15 de enero de 2020 13:30:45 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - HOUR_ONLY 1`] = `"13"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_DATE 1`] = `"miércoles, 15 de enero de 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_DATE_WITH_YEAR 1`] = `"miércoles, 15 de enero de 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_TIME 1`] = `"13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_TIME_WITH_TIMEZONE 1`] = `"13:30:45 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_WEEKDAY 1`] = `"miércoles"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_WEEKDAY_LONG_TIME 1`] = `"miércoles, 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_WEEKDAY_SHORT_TIME 1`] = `"miércoles, 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_WITH_TIMEZONE 1`] = `"miércoles, 15 de enero de 2020 13:30:45 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - LONG_WITH_YEAR_TIMEZONE 1`] = `"miércoles, 15 de enero de 2020 13:30:45 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - MEDIUM 1`] = `"miércoles, 15 de enero de 2020 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - MEDIUM_DATE 1`] = `"miércoles, 15 de enero de 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - MEDIUM_DATE_SHORT_TIME 1`] = `"15 ene, 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - MEDIUM_DATE_WITH_YEAR 1`] = `"miércoles, 15 de enero de 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - MEDIUM_TIME 1`] = `"13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - MEDIUM_WITH_YEAR 1`] = `"miércoles, 15 de enero de 2020 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT 1`] = `"15/01/2020 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_DATE 1`] = `"15/01/2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_DATE_LONG_TIME 1`] = `"15/01/2020 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_DATE_TIME 1`] = `"15/01/2020 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_DATE_WITH_SHORT_YEAR 1`] = `"15/01/2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_DATE_WITH_YEAR 1`] = `"15/01/2020"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_TIME 1`] = `"13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_WEEKDAY 1`] = `"mié"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_WEEKDAY_LONG_TIME 1`] = `"mié, 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_WEEKDAY_SHORT_TIME 1`] = `"mié, 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting es-es - SHORT_WITH_YEAR 1`] = `"15/01/2020 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - FULL 1`] = `"miercuri, 15 ianuarie 2020 13:30:45 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - FULL_DATE 1`] = `"miercuri, 15 ianuarie 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - FULL_DATE_WITH_YEAR 1`] = `"miercuri, 15 ianuarie 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - FULL_TIME 1`] = `"13:30:45 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - FULL_WITH_YEAR 1`] = `"miercuri, 15 ianuarie 2020 13:30:45 GMT+00:00"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - HOUR_ONLY 1`] = `"13"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_DATE 1`] = `"miercuri, 15 ianuarie 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_DATE_WITH_YEAR 1`] = `"miercuri, 15 ianuarie 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_TIME 1`] = `"13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_TIME_WITH_TIMEZONE 1`] = `"13:30:45 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_WEEKDAY 1`] = `"miercuri"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_WEEKDAY_LONG_TIME 1`] = `"miercuri, 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_WEEKDAY_SHORT_TIME 1`] = `"miercuri, 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_WITH_TIMEZONE 1`] = `"miercuri, 15 ianuarie 2020 13:30:45 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - LONG_WITH_YEAR_TIMEZONE 1`] = `"miercuri, 15 ianuarie 2020 13:30:45 GMT+0"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - MEDIUM 1`] = `"miercuri, 15 ianuarie 2020 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - MEDIUM_DATE 1`] = `"miercuri, 15 ianuarie 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - MEDIUM_DATE_SHORT_TIME 1`] = `"15 ian., 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - MEDIUM_DATE_WITH_YEAR 1`] = `"miercuri, 15 ianuarie 2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - MEDIUM_TIME 1`] = `"13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - MEDIUM_WITH_YEAR 1`] = `"miercuri, 15 ianuarie 2020 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT 1`] = `"15.01.2020 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_DATE 1`] = `"15.01.2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_DATE_LONG_TIME 1`] = `"15.01.2020 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_DATE_TIME 1`] = `"15.01.2020 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_DATE_WITH_SHORT_YEAR 1`] = `"15.01.2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_DATE_WITH_YEAR 1`] = `"15.01.2020"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_TIME 1`] = `"13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_WEEKDAY 1`] = `"mie."`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_WEEKDAY_LONG_TIME 1`] = `"mie., 13:30:45"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_WEEKDAY_SHORT_TIME 1`] = `"mie., 13:30"`;

exports[`Windows Regional Formats > Regional settings based formatting ro-ro - SHORT_WITH_YEAR 1`] = `"15.01.2020 13:30"`;
